go/vt/vtgate/buffer/shard_buffer.go

```diff
@@ -182,28 +200,109 @@ func (sb *shardBuffer) logErrorIfStateNotLocked(state bufferState) {
-		return nil, vterrors.FromError(vtrpcpb.ErrorCode_TRANSIENT_ERROR, errors.New("master buffer is full"))
-	}
+		// Buffer is full. Evict the oldest entry and buffer this request instead.
+		if len(sb.queue) == 0 {
+			// Overall buffer is full, but this shard's queue is empty. That means
+			// there is at least one other shard failing over as well which consumes
+			// the whole buffer.
+			return nil, bufferFullError
+		}
 
-	// TODO(mberlin): Kill the oldest entry if full.
+		e := sb.queue[0]
+		// Evict the entry. Do not release its slot in the buffer and reuse it for
+		// this new request.
+		// NOTE: We keep the lock to avoid racing with drain().
+		sb.unblockAndWait(e, entryEvictedError, false /* releaseSlot */)
+		sb.queue = sb.queue[1:]
+	}
 
 	e := &entry{
-		done: make(chan struct{}),
+		done:     make(chan struct{}),
+		deadline: time.Now().Add(*window),
 	}
 	e.bufferCtx, e.bufferCancel = context.WithCancel(ctx)
 	sb.queue = append(sb.queue, e)
 	requestsInFlightMax.Add(sb.statsKey, 1)
+	if len(sb.queue) == 1 {
+		sb.timeoutThread.notifyQueueNotEmpty()
+	}
 	return e, nil
 }
 
+// unblockAndWait unblocks a blocked request and waits until it reported its end.
+func (sb *shardBuffer) unblockAndWait(e *entry, err error, releaseSlot bool) {
+	// Set error such that the request will see it.
+	e.err = err
+	// Tell blocked request to stop waiting.
+	close(e.done)
+	// Wait for unblocked request to end.
+	<-e.bufferCtx.Done()
+	if releaseSlot {
+		sb.bufferSizeSema.Release()
+	}
+}
+
 // wait blocks while the request is buffered during the failover.
 func (sb *shardBuffer) wait(ctx context.Context, e *entry) error {
 	select {
 	case <-ctx.Done():
-		// TODO(mberlin): Cancel buffering. Implement an sb.cancel(e) for that.
+		sb.remove(e)
 		return vterrors.FromError(vtrpcpb.ErrorCode_TRANSIENT_ERROR, fmt.Errorf("context was canceled before failover finished (%v)", ctx.Err()))
 	case <-e.done:
-		return nil
+		return e.err
+	}
+}
+
+// oldestEntry returns the head of the queue or nil if the queue is empty.
+func (sb *shardBuffer) oldestEntry() *entry {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+
+	if len(sb.queue) > 0 {
+		return sb.queue[0]
+	}
+	return nil
+}
+
+// evictOldestEntry is used by timeoutThread to evict the head entry of the
+// queue if it exceeded its buffering window.
+func (sb *shardBuffer) evictOldestEntry(e *entry) {
+	sb.mu.Lock()
+	if len(sb.queue) == 0 || e != sb.queue[0] {
+		// Entry is already removed e.g. by remove(). Ignore it.
+		return
+	}
+	sb.unblockAndWait(e, nil /* err */, true /* releaseSlot */)
+	sb.queue = sb.queue[1:]
+	sb.mu.Unlock()
+}
+
+// remove must be called when the request was canceled from outside and not
+// internally.
+func (sb *shardBuffer) remove(toRemove *entry) {
+	sb.mu.Lock()
+	defer sb.mu.Unlock()
+
+	if sb.queue == nil {
+		// Queue is cleared because we're already in the DRAIN phase.
+		return
+	}
+
+	// Use the same array to avoid an allocation. Copy all but "toRemove".
```

Comment:
This is not obvious to me. I think the first append will not reuse the same underlying buffer, but re-allocate.

The more usual way to do this is to:
for i, e := range sb.queue {
  if e == toRemove {
    copy(sb.queue[i:], sb.queue[i+1:])
    break
  }
}
or whatever the right indexes are, I forgot. copy() says it supports this.